OpType.CnZ only accepts >0 control qubits but should accept >=0

Currently fails tests at pytket-qiskit

```
FAILED qiskit_convert_test.py::test_convert_cnz_to_qiskit - qiskit.circuit.exceptions.CircuitError: 'The number of control qubits must be in `[1, num_qubits - base_gate.num_qubits]`.'

--- a/pytket/conanfile.py
+++ b/pytket/conanfile.py
@@ -37,7 +37,7 @@ class pytketRecipe(ConanFile):
         self.requires("nanobind/tci-2.7.0@tket/stable")
         self.requires("symengine/tci-0.14.0@tket/stable")
         self.requires("tkassert/0.3.4@tket/stable")
-        self.requires("tket/2.1.10@tket/stable")
+        self.requires("tket/2.1.11@tket/stable")
         self.requires("tklog/0.3.3@tket/stable")
         self.requires("tkrng/0.3.3@tket/stable")
         self.requires("tktokenswap/0.3.11@tket/stable")

--- a/pytket/docs/changelog.rst
+++ b/pytket/docs/changelog.rst
@@ -1,6 +1,14 @@
 Changelog
 =========
 
+Unreleased
+----------
+
+Fixes:
+
+* Apply special handling of multi-controlled gates with 0 or 1 controls when
+  adding operations using `UnitID` arguments.
+
 2.3.1 (April 2025)
 ------------------
 

--- a/pytket/tests/circuit_test.py
+++ b/pytket/tests/circuit_test.py
@@ -1634,6 +1634,15 @@ def test_cnx_vchain_arbitrary_ancillas() -> None:
         assert np.allclose(m, m_cnx_with_ancillas)
 
 
+def test_c0z() -> None:
+    # https://github.com/CQCL/tket/issues/1873
+    circ = Circuit(1).add_gate(OpType.CnZ, [0])
+    cmds = circ.get_commands()
+    assert len(cmds) == 1
+    cmd = cmds[0]
+    assert cmd.op.type == OpType.Z
+
+
 if __name__ == "__main__":
     test_circuit_gen()
     test_symbolic_ops()

--- a/tket/conanfile.py
+++ b/tket/conanfile.py
@@ -23,7 +23,7 @@ from conan.errors import ConanInvalidConfiguration
 
 class TketConan(ConanFile):
     name = "tket"
-    version = "2.1.10"
+    version = "2.1.11"
     package_type = "library"
     license = "Apache 2"
     homepage = "https://github.com/CQCL/tket"

--- a/tket/include/tket/Circuit/Circuit.hpp
+++ b/tket/include/tket/Circuit/Circuit.hpp
@@ -46,6 +46,7 @@
 #include "DAGDefs.hpp"
 #include "ResourceData.hpp"
 #include "Slices.hpp"
+#include "tket/Gate/GatePtr.hpp"
 #include "tket/Gate/OpPtrFunctions.hpp"
 #include "tket/Utils/Json.hpp"
 #include "tket/Utils/SequencedContainers.hpp"
@@ -1739,7 +1740,26 @@ Vertex Circuit::add_op(
     preds.push_back(pred_out_e);
   }
 
-  Vertex new_v = add_vertex(op, opgroup);
+  Vertex new_v;
+  OpType optype = op->get_type();
+  if (optype == OpType::CnRy && args.size() == 1) {
+    new_v = add_vertex(
+        get_op_ptr(OpType::Ry, as_gate_ptr(op)->get_params()), opgroup);
+  } else if (optype == OpType::CnRx && args.size() == 1) {
+    new_v = add_vertex(
+        get_op_ptr(OpType::Rx, as_gate_ptr(op)->get_params()), opgroup);
+  } else if (optype == OpType::CnRz && args.size() == 1) {
+    new_v = add_vertex(
+        get_op_ptr(OpType::Rz, as_gate_ptr(op)->get_params()), opgroup);
+  } else if (optype == OpType::CnX && args.size() == 1) {
+    new_v = add_vertex(get_op_ptr(OpType::X), opgroup);
+  } else if (optype == OpType::CnZ && args.size() == 1) {
+    new_v = add_vertex(get_op_ptr(OpType::Z), opgroup);
+  } else if (optype == OpType::CnY && args.size() == 1) {
+    new_v = add_vertex(get_op_ptr(OpType::Y), opgroup);
+  } else {
+    new_v = add_vertex(op, opgroup);
+  }
   rewire(new_v, preds, sig);
   return new_v;
 }

--- a/tket/src/Circuit/basic_circ_manip.cpp
+++ b/tket/src/Circuit/basic_circ_manip.cpp
@@ -121,7 +121,6 @@ Vertex Circuit::add_op<unsigned>(
         gate->get_name() + " requires " + std::to_string(sig.size()));
   }
 
-  OpType optype = gate->get_type();
   unit_vector_t arg_ids;
   for (unsigned i = 0; i < args.size(); ++i) {
     switch (sig.at(i)) {
@@ -138,19 +137,6 @@ Vertex Circuit::add_op<unsigned>(
         TKET_ASSERT(!"add_op found invalid edge type in signature");
       }
     }
-  }
-  if (optype == OpType::CnRy && args.size() == 1) {
-    return add_op(get_op_ptr(OpType::Ry, gate->get_params()), arg_ids);
-  } else if (optype == OpType::CnRx && args.size() == 1) {
-    return add_op(get_op_ptr(OpType::Rx, gate->get_params()), arg_ids);
-  } else if (optype == OpType::CnRz && args.size() == 1) {
-    return add_op(get_op_ptr(OpType::Rz, gate->get_params()), arg_ids);
-  } else if (optype == OpType::CnX && args.size() == 1) {
-    return add_op(get_op_ptr(OpType::X), arg_ids);
-  } else if (optype == OpType::CnZ && args.size() == 1) {
-    return add_op(get_op_ptr(OpType::Z), arg_ids);
-  } else if (optype == OpType::CnY && args.size() == 1) {
-    return add_op(get_op_ptr(OpType::Y), arg_ids);
   }
   return add_op(gate, arg_ids, opgroup);
 }

